csrc/net/ncnn/ncnn_net.cpp

```diff
@@ -91,11 +111,25 @@ Result<void> NCNNNet::Forward() {
     OUTCOME_TRY(ncnn_status(extractor.extract(output_indices_[i], outputs[i])));
     auto& tensor = output_tensors_[i];
     auto shape = outputs[i].shape();
-    tensor.Reshape({1, shape.w, shape.h, shape.c});
+    if (outputs[i].dims == 1) {
+      tensor.Reshape({1, shape.w});
+    } else if (outputs[i].dims == 2) {
+      tensor.Reshape({1, shape.h, shape.w});
+    } else if (outputs[i].dims == 3) {
+      tensor.Reshape({1, shape.d, shape.h, shape.w});
+    } else {
+      // for dim==4 case and blank image.
+      tensor.Reshape({1, shape.c, shape.d, shape.h, shape.w});
+    }
+    // tensor.Reshape({1, shape.c, shape.h, shape.w});
     // ncnn Mat may be padded, flatten to avoid that
-    auto flattened = outputs[i].reshape(shape.w * shape.h * shape.c);
-    OUTCOME_TRY(tensor.CopyFrom(flattened.data, stream_));
+    auto flattened = outputs[i].reshape(shape.c * shape.h * shape.w);
+    // if ((shape.c * shape.h * shape.w) > 0)
+    if (outputs[i].dims > 0) {
+      OUTCOME_TRY(tensor.CopyFrom(flattened.data, stream_));
+    }
   }
+  OUTCOME_TRY(stream_.Wait());
```

Comment:
There is a long standing use-after-free bug here. When reshaping really happens (the `Mat` is padded), `flattened` become the only owner of it's data; soon it goes out of scope and free it's data before the asynchronous copy complete.

To fix it we have to either wait for every copy or use a `std::vector` from outer scope to hold every `flattened`.